agatha/construct/semrep_util.py

```diff
@@ -302,15 +318,101 @@ def _get_flags(self, input_path:Path, output_path:Path)->List[str]:
     if self.word_sense_disambiguation:
       res.append("--word_sense_disambiguation")
     res.append("--xml_output_format")
-    res.append(str(input_path))
-    res.append(str(output_path))
+    if input_path is not None and output path is not None:
+      res.append(str(input_path))
+      res.append(str(output_path))
+    else:
+      raise ValueError("Both input and output paths must be set or None")
     return res
 
-  def run(self, input_path:Path, output_path:Path)->None:
+  def _read_lines(self, in_file:Path)->List[str]:
+    """
+    Reads in lines of a file in memory.
+
+    Args:
+     in_file: Path to input file
+    Returns:
+      lines in the files
+    """
+    with open(in_file, 'r') as f:
+      lines = []
+      for line in f:
+        lines.append(line.strip())
+    return lines
+
+  def _write_xml(self, data: List[lxml.etree._Element], out_file:Path)->None:
+    """
+    Writes the produced XML to the specified
+    file.
+
+    Args:
+      data: List of xml elements produced by semrep
+      out_file: Path to output xml file
+    """
+    assert out_file.suffix == '.xml', 'The output must be a valid XML file'
+    with open(out_file, 'w') as ofile:
+      ofile.write(
+          '\n'.join(
+            xml
+            for xml in data
+            )
+          )
+
+
+
+  def _run_gracefully(self,
+      in_lines:List[str],
+      )->List[lxml.etree._Element]:
+    """
+    Runs individual lines of the text through semrep and
+    produces the output xml. Raises a `ChildProcessError`
+    that is caught in the actual `run()` function.
+
+    The input is expected to be a list of strings of the form:
+
+    `pmid|text`
+
+    Args:
+      in_lines: Preprocessed semrep text.
+
+    Returns:
+      List of XML elements if no exception is raised
+    """
+    cmd = self._get_flags(None, None)
+    env = self._get_env()
+    semrep_proc = subprocess.Popen(
+        cmd,
+        env=env,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE
+        )
+
+    out_lines, error = semrep_proc.communicate(
+        "\n".join(
+          in_lines
+          )
+        )
+    if semrep_proc.returncode != 0:
+      raise ChildProcessError(
+          "Semrep failed with code:{} and error:{}".format(
+            semrep_proc.returncode,
+            error
+            )
+          )
+    return out_lines
+
+
+
+  def run(self,
+      input_path:Path,
+      output_path:Path,
+      line_mode:bool=True
+      )->None:
     """Actually calls SemRep with an input file.
 
     Args:
       input_path: The location of the SemRep Input file
+      line_mode: Flag to run semrep on files or lines.
```

Comment:
A convention regarding boolean input parameters. Its preferred to have all boolean params default to `False` because then when you call the function you can "enable" different behaviors. Additionally, the parameter names should sound like verbs, so its a little clearer what is being enabled. For this reason I think it would be more readable to change `line_mode:bool=True` to `use_file_interface:bool=False` or something similar. For this reason, if we ever got to the situation where we wanted the non-standard behavior of this function it would be a little clearer what was happening.